File: cblmapper/src/main/java/io/sportner/cblmapper/CBLDocumentChangeListener.java
package io.sportner.cblmapper;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import com.couchbase.lite.Database;
import com.couchbase.lite.DocumentChange;
import com.couchbase.lite.DocumentChangeListener;
import io.sportner.cblmapper.exceptions.CBLMapperClassException;

public abstract class CBLDocumentChangeListener<T extends CBLDocument> implements DocumentChangeListener {
    private final CBLMapper mCBLMapper;
    private final Database mDatabase;
    private final Class<T> mType;

    public CBLDocumentChangeListener(@NonNull Database database, @NonNull CBLMapper mapper, @NonNull Class<T> type) {
        this.mDatabase = database;
        this.mCBLMapper = mapper;
        this.mType = type;
    }

    public void changed(DocumentChange change) {
        try {
            this.onDocumentChange(this.mCBLMapper.fromDocument(this.mDatabase.getDocument(change.getDocumentID()), this.mType));
        } catch (CBLMapperClassException var3) {
            var3.printStackTrace();
        }

    }

    public abstract void onDocumentChange(@Nullable T var1);
}

File: cblmapper/src/main/java/io/sportner/cblmapper/CBLDocumentChangeListener.kt
package io.sportner.cblmapper

import com.couchbase.lite.Database
import com.couchbase.lite.DocumentChange
import com.couchbase.lite.DocumentChangeListener
import io.sportner.cblmapper.exceptions.CBLMapperClassException

abstract class CBLDocumentChangeListener<T : CBLDocument>(private val mDatabase: Database, private val mCBLMapper: CBLMapper, private val mType: Class<T>) : DocumentChangeListener {

    override fun changed(change: DocumentChange) {
        try {
            this.onDocumentChange(this.mCBLMapper.fromDocument(this.mDatabase.getDocument(change.documentID), this.mType))
        } catch (var3: CBLMapperClassException) {
            var3.printStackTrace()
        }

    }

    abstract fun onDocumentChange(var1: T?)
}
